### ISSUE TITLE
Final Term command marks / shell integration

### ISSUE THREAD

**Issue by krzysdz** [JanDeDobbeleer/oh-my-posh#3795]
### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### What would you like to see changed/added?

# Final Term shell integration

Oh My Posh currently supports the iTerm prompt mark through a template. This is, however, supported only in iTerm with bash and zsh.

A similar functionality can be achieved using Final Term's control sequences. These sequences allow the terminal to get more information about the shell contents - the exact locations of commands and their outputs, as well as the command exit codes. iTerm2, kitty, Windows Terminal and the VS Code integrated terminal can use `FTCS` sequences for better shell integration and showing marks next to commands.

There are 4 sequences that are required for the integration:

1. `FTCS_PROMPT` - `OSC 133; A ST`: marks the start of the prompt,
2. `FTCS_COMMAND_START` - `OSC 133; B ST`: marks the start of the command,
3. `FTCS_COMMAND_EXECUTED` - `OSC 133; C ST`: marks the start of the command output,
4. `FTCS_COMMAND_FINISHED` - `OSC 133; D [; <code> ] ST`: marks the end of the output and sends the exit code (if a command has been executed).

## Examples

CMD and pwsh example prompts (without `FTCS_COMMAND_EXECUTED`): microsoft/terminal#14341

### Shell integration in terminal emulators

Example integration in Windows Terminal (with some experimental settings turned on):
![Windows Terminal shell integration example with iTerm2's FTCS support code in background](https://user-images.githubusercontent.com/12915102/236363515-982af391-c4cc-4fda-bc76-e20e8884b6be.png)

Example integration in VS Code:
![Visual Studio Code shell integration example](https://user-images.githubusercontent.com/12915102/236364303-bd914e4f-f1ba-4cc4-8818-32d4b317894c.png)

Test string (pwsh):
```powershell
echo "`n`e]133;A`a> `e]133;B`aa command`n`e]133;C`amultiline`noutput`naaaa`n`e]133;D`a`e]133;A`a> `e]133;B`aanother command`n`e]133;C`aoutput`n`e]133;D;1`a`n`e]133;A`aprompt decoration`n> `e]133;B`aa \`n> multiline \`n> command`n`e]133;C`awith`na`nlong`noutput`n`e]133;D`a"
```

## More information

- Terminal Working Group discussion on line marks: https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/28
- iTerm2's escape codes documentation: https://iterm2.com/documentation-escape-codes.html
- Windows Terminal scroll marks thread: microsoft/terminal#11000
- kitty documentation and example zsh, fish and bash integration code: https://sw.kovidgoyal.net/kitty/shell-integration

**Comment by JanDeDobbeleer** [comment]
@krzysdz if we do this, it needs to be added to the engine and not in each shell script as that's cumbersome to maintain. It can even have an easy config setting to enable this that way.

**Comment by JanDeDobbeleer** [comment]
@krzysdz had a quick look and it seems that only `FTCS_COMMAND_EXECUTED` is annoying from the engine as we'll need to leverage that from the shell itself (like we do for transient). This implies that's limited to shell's offering that [hook](https://ohmyposh.dev/docs/configuration/transient).

**Comment by JanDeDobbeleer** [comment]
@krzysdz I have a candidate ready for this but which setting is this?

> Example integration in Windows Terminal (with some experimental settings turned on):

**Comment by krzysdz** [comment]
`"experimental.showMarksOnScrollbar": true` for marks on the scroll bar and the right click menu with *select command/output* is `"experimental.rightClickContextMenu": true`. Both settings are per-profile.

**Comment by JanDeDobbeleer** [comment]
@krzysdz tried that, version is correct but doesn't work. I assumed I can add that to the default profile setting.

**Comment by krzysdz** [comment]
Looks like it doesn't work in the release version of Windows Terminal yet - only Preview and Dev.

Here's a screenshots from my own experiments:
![FTCS shell integration demo in different Windows Terminal verisons](https://github.com/JanDeDobbeleer/oh-my-posh/assets/12915102/16d52a72-e1d7-439f-a126-ffad6c106d73)

**Comment by JanDeDobbeleer** [comment]
@krzysdz I'm on preview. I'll check what I did wrong.

**Comment by krzysdz** [comment]
Shells in the screenshot use [this version](https://github.com/krzysdz/oh-my-posh/tree/shell-integration), which sometimes works (good enough for me in PowerShell), but is very limited and broken in some shells.

**Comment by JanDeDobbeleer** [comment]
@krzysdz I have a very similar approach, but probably I did something wrong on my end so I'll compare and adjust. Awesome.

**Comment by JanDeDobbeleer** [comment]
@krzysdz I have the scroll marks, but how do I get a context menu? That doesn't work on my end.

**Comment by JanDeDobbeleer** [comment]
@krzysdz do you mind creating a PR with your changes? I'd love to tweak that a little bit and integrate it.

**Comment by krzysdz** [comment]
> @krzysdz I have the scroll marks, but how do I get a context menu? That doesn't work on my end.

I was merged in microsoft/terminal#15020 a while ago, but I don't know if it already is in the Preview version.

**Comment by krzysdz** [comment]
> @krzysdz do you mind creating a PR with your changes? I'd love to tweak that a little bit and integrate it.

Sure, I'll do it later today.
